src/compiler/checker.ts

```diff
@@ -36699,9 +36700,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const typeArgumentNodes: readonly TypeNode[] | undefined = callLikeExpressionMayHaveTypeArguments(node) ? node.typeArguments : undefined;
         const instantiated = typeArgumentNodes
             ? createSignatureInstantiation(candidate, getTypeArgumentsFromNodes(typeArgumentNodes, typeParameters, isInJSFile(node)))
+            // when there is only one candidate reuse existing *inferred* candidate for argument error (if available)
+            // this saves the compiler some extra work but more importantly it includes inferences made from context-sensitive arguments and generic functions
+            // which avoids confusing mismatches between inferred type arguments and reported argument error
+            //
+            // for the time being only do this when there is a single candidate as the origin index of the `candidatesForArgumentError` is not known
+            // and it could be different from the selected `bestIndex` here

```

Comment:
I'd consider revisiting this in Corsa where `CallState` is already inroduced and improving upon this would likely be slightly easier thanks to that